content/tutorials/esc-managed-chatapp/index.md

```diff
@@ -0,0 +1,533 @@
+
+After we get a basic version of the app running, we'll modify the app two times, to show three different ways of managing our secrets and configurations; unmanaged, managed static configuration, and managed dynamic configuration.
+
+The first version of our app will be *unmanaged* and will have hardcoded settings for the OpenAI model we're using, reading it from the default environment variable `OPENAI_API_KEY` for the API key.
+
+The second version will be *statically managed* using Pulumi ESC, providing a custom set of environment variables to the app at runtime.
+
+The third version will be *dynamically managed* via the Pulumi ESC SDK to reload the configuration, so that there's no downtime required to reconfigure the web app.
+
+{{< figure src="chatapp-configuration-versions.svg" caption="Figure: Three different versions of the code, using different configuration methods">}}
+
+## Creating the ChatApp
+
+{{% notes type="info" %}}
+**Before you start:** Make sure you have an [OpenAI account](https://platform.openai.com/docs/quickstart) with enough credits to test your app.
+{{% /notes %}}
+
+### Step 1: Create a OpenAI project, and service account API key
+
+First we'll need to create an OpenAI project called `chatapp`.
+
+{{< figure src="openai-create-project-dialog.png" caption="Figure: OpenAI Create Project dialog">}}
+
+Next, create a service account API key. You can follow the [OpenAI Quickstart](https://platform.openai.com/docs/quickstart) docs for details.
+
+{{< figure src="openai-create-secret-dialog.png" caption="Figure: OpenAI Create API Key dialog.">}}
+
+Next, copy the key to your clipboard, and use the following command to set the `OPENAI_API_KEY` environment variable:
+
+```bash
+$ export OPENAI_API_KEY=`pbpaste`
+```
+
+{{% notes type="warning" %}}
+**Shell environments may differ:** This tutorial is assuming a Bash-like shell in a MacOS environment where the `pbcopy` and `pbpaste` tools are available by default.
+
+You may need to install a similar tool for your environment to avoid pasting the API key text in plain-text into your shell (and by proxy, into your shell command history).
+
+Two such utilities are `xclip` and `xsel`, which are available in the default repositories of most Linux distributions.
+{{% /notes %}}
+
+### Step 2: Install dependencies with pip.
+
+Now lets setup our working environment. Create a new directory called `chatapp`.
+
+```bash
+$ mkdir chatapp
+$ cd chatapp
+```
+
+Next create a Python virtual environment:
+
+```bash
+$ python -m venv .venv
+```
+
+And then install the `flask` and `openai` packages
+
+```bash
+pip install flask openai
+```
+
+### Step 3: Create the webapp
+
+Finally, lets create the code files for our chat app. We'll need two files `chatapp.py`, the backend Flask app, and `templates/chatapp.html`, an HTML template that will render the UI. You can copy and paste these from here, into your favorite code editor and save.
+
+**Code Listing:** `chatapp.py`
+
+```python
```

Comment:
might consider adding something to give an error if they don't correctly get their API key in: 

import os
from flask import Flask, render_template, request, jsonify
from openai import OpenAI

app = Flask(__name__)

def get_openai_response(api_key, model, prompt):
    client = OpenAI(api_key=api_key)
    try:
        response = client.chat.completions.create(
            model=model,
            messages=[{"role": "user", "content": prompt}]
        )
        # Print the full response for debugging
        print(response)
        # Return only the content of the message
        return response.choices[0].message.content
    except Exception as e:
        return f"An error occurred: {str(e)}"

@app.route('/')
def home():
    return render_template('chatapp.html')

@app.route('/chat', methods=['POST'])
def chat():
    model = "gpt-3.5-turbo"
    api_key = os.getenv("OPENAI_API_KEY")

    prompt = request.json['message']
    response = get_openai_response(api_key, model, prompt)
    return jsonify({'response': response, 'model': model})

if __name__ == '__main__':
    app.run(debug=True)